### PR TITLE
ci: add realase candidate pipeline action [SFUI2-1184]

### PR DESCRIPTION
**Pull request by Szymon-dziewonski**
# Related issue

https://vsf.atlassian.net/browse/SFUI2-1184

# Scope of work

Added 3 github actions
- Create changelog PR from changeset on every commit push into `v2-release/**` 
- Create Trigger if this changelog should be for release candidate (rc tag) or production (latest)
- If latest commit message in `v2-release/**` is `[ci] release rc` should run publish with tag `rc`

File `release-changeset-changelog.yml` job `create-release-changelog`
Run every time commit lands into `v2-release/**` branch and creates PR with new changelog either for RC or PROD changelog

File `ci.yml` job `release-next`
Runs only on `v2-release` when last pushed commit name is `[ci] release next` and publish release-candidate with tag `next`, this is benefitial because we can run publish after all `lint` `build` `vue-test` `react-test` checks

File `workflow-dispatch-changeset.yml` job `generation-changelog-on-trigger`
Run after manual trigger regeneration and pushing commit with `.changelog/pre.json` 

image

# Screenshots of visual changes

# Checklist

- [x] Self code-reviewed
- [ ] Changes documented
- [ ] Semantic HTML
- [ ] SSR-friendly
- [ ] Caching friendly
- [ ] a11y for WCAG 2.0 AA
- [ ] examples created
- [ ] blocks created
- [ ] cypress tests created

### CODE REVIEW

File: .github/workflows/release-changeset-changelog.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  push:
+    branches:
+      - v2-release/**
+
+jobs:
+  release-canary:
+    name: Release canary package
+    needs: [build, build-release, cypress-react, cypress-vue]
+    if: ${{ github.ref == 'refs/heads/v2-develop' }}
+    # Can't use 'configure-enviroment' because it fails on dev branch
```
Comment (FRSgit) [inline .github/workflows/release-changeset-changelog.yml:12]: Do you know why it fails?

File: .github/workflows/release-changeset-changelog.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
+name: Generate changelog on v2-release
+on:
+  push:
+    branches:
+      - v2-release/**
```
Comment (FRSgit) [inline .github/workflows/release-changeset-changelog.yml:5]: Suggestion:
This should be `v2-develop`, right?

File: .github/workflows/workflow-dispatch-changeset.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        uses: actions/checkout@v3
+      - name: Install Node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version-file: '.node-version'
+          cache: 'yarn'
+      - name: Install dependencies
+        run: yarn --immutable
+      - name: Change changeset pre based on trigger
+        run: yarn changeset pre ${{ inputs.generate-changelog }} ${{ inputs.generate-changelog == 'enter' && 'next' }}
```
Comment (FRSgit) [inline .github/workflows/workflow-dispatch-changeset.yml:28]:
```suggestion
        run: yarn changeset pre ${{ inputs.generate-changelog }} ${{ inputs.generate-changelog == 'enter' && 'rc' }}
```

File: .github/workflows/workflow-dispatch-changeset.yml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      - name: Create PR with changelog
+        id: changesets
+        uses: changesets/action@v1
+        with:
+          version: yarn run changeset-version
+          commit: '[ci] - release ${{ inputs.generate-changelog }}'
```
Comment (FRSgit) [inline .github/workflows/workflow-dispatch-changeset.yml:34]: `inputs.generate-changelog` equals to enter, right?

```suggestion
          commit: '[ci] - release rc'
```

File: .github/workflows/workflow-dispatch-changeset.yml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        uses: actions/setup-node@v3
+        with:
+          node-version-file: '.node-version'
+          cache: 'yarn'
+      - name: Install dependencies
+        run: yarn --immutable
+      - name: Change changeset pre based on trigger
+        run: yarn changeset pre ${{ inputs.generate-changelog }} ${{ inputs.generate-changelog == 'enter' && 'next' }}
+      - name: Create PR with changelog
+        id: changesets
```
Comment (FRSgit) [inline .github/workflows/workflow-dispatch-changeset.yml:30]:
```suggestion
        id: changesets
        if: ${{ inputs.generate-changelog == 'enter' }}
```

File: .github/workflows/workflow-dispatch-changeset.yml (hunk 4 above)
Comment (FRSgit) [reply .github/workflows/workflow-dispatch-changeset.yml:34]: Btw. where does it get commited into?

File: .github/workflows/release-changeset-changelog.yml (hunk 1 above)
Comment (Szymon-dziewonski) [reply .github/workflows/release-changeset-changelog.yml]:
```
          else
            exit 1
          fi
```

Comment (Szymon-dziewonski) [reply .github/workflows/release-changeset-changelog.yml]: but it doesnt matter, I pushed old version, now this whole file looks different

File: .github/workflows/workflow-dispatch-changeset.yml (hunk 5 above)
Comment (Szymon-dziewonski) [reply .github/workflows/workflow-dispatch-changeset.yml:30]: it should run always, no matter if its prod or next

File: .github/workflows/workflow-dispatch-changeset.yml (hunk 3 above)
Comment (Szymon-dziewonski) [reply .github/workflows/workflow-dispatch-changeset.yml:28]: I thought we established that we want to release rc with tag `next`

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -230,3 +231,12 @@ jobs:
           # refresh token before Saturday, May 25, 2024
           NPM_TOKEN: ${{ secrets.NPM_RELEASE_TOKEN }}
           NODE_AUTH_TOKEN: ${{ secrets.NPM_RELEASE_TOKEN }}
+
+  release-next:
+    name: Release next release candidate
+    runs-on: ubuntu-latest
+    needs: [build, build-release, cypress-react, cypress-vue]
+    if: ${{ startsWith( github.ref, 'refs/heads/v2-release' ) && github.event.head_commit.message == '[ci] - release next' }}
```
[the remark this answers is not in the document]
Comment (Szymon-dziewonski) [reply .github/workflows/ci.yml:240]: added

File: .github/workflows/release-changeset-changelog.yml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        uses: actions/checkout@v3
+      - name: Install Node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version-file: '.node-version'
+          cache: 'yarn'
+      - name: Install dependencies
+        run: yarn --immutable
+      - name: ConsoleLOG
+        ron: echo fromJSON($(cat .changeset/pre.json )).mode
```
Comment (FRSgit) [inline .github/workflows/release-changeset-changelog.yml:23]:
```suggestion
        run: echo fromJSON($(cat .changeset/pre.json )).mode
```

File: .github/workflows/release-changeset-changelog.yml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      - name: Set VAR if .changeset/pre.json file contain "mode:pre"
+        id: vars
+        run: echo "isInPreMode=$(echo ${{ fromJSON($(cat .changeset/pre.json )).mode == "pre" }})" >> $GITHUB_OUTPUT
```
Comment (FRSgit) [inline .github/workflows/release-changeset-changelog.yml:26]:
```suggestion
      - name: Check if .changeset/pre.json file contain "mode:pre"
        id: prerelease_check
        run: echo "result=$(echo ${{ fromJSON($(cat .changeset/pre.json )).mode == "pre" }})" >> $GITHUB_OUTPUT
```

Comment (FRSgit) [reply .github/workflows/release-changeset-changelog.yml:26]:
```suggestion
      - name: Set VAR if .changeset/pre.json file contain "mode:pre"
        id: vars
        run: |
    if ${{ fromJSON($(cat .changeset/pre.json )).mode == "pre" }}; then
      echo "result=true" >> $GITHUB_OUTPUT
    else
      echo "result=false" >> $GITHUB_OUTPUT
    fi
```

**Review by FRSgit — APPROVED** [review]

**Pull request merged**